### Initial Analysis

In [4]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns



1. Load the Foreign Gifts data.
     2. Describe the differences between different classes of gift types.
     3. Answer the following questions:
         1. Which country gives the most money in total?
         2. Which country initiates the most gifts by count?
         3. Which country gives the largest gifts on average?
         4. Which institution receive the most money in total?
         5. Which institution receive the greatest number of gifts by count?
         6. What is the average amount each university receives? Median amount? If they are different why?
         7. What is the largest flow of a single country to a single institution?
         8. Which are the top giftors to US academic institutions?

### Load the Foreign Gifts data.

In [5]:
fgData = pd.read_csv('ForeignGifts_edu.csv')

# fgData.head()
#create a copy of fgData to work with
df = fgData.copy()

fgData

,ID,OPEID,Institution Name,City,State,Foreign Gift Received Date,Foreign Gift Amount,Gift Type,Country of Giftor,Giftor Name
0,1,102000,Jacksonville State University,Jacksonville,AL,43738,250000,Monetary Gift,CHINA,NaN
1,2,104700,Troy University,Troy,AL,43592,463657,Contract,CHINA,Confucius Institute Headquarters
2,3,105100,University of Alabama,Tuscaloosa,AL,43466,3649107,Contract,ENGLAND,Springer Nature Customer Service Ce
3,4,105100,University of Alabama,Tuscaloosa,AL,43472,1000,Contract,SAUDI ARABIA,Saudi Arabia Education Mission
4,5,105100,University of Alabama,Tuscaloosa,AL,43479,49476,Contract,SAUDI ARABIA,Saudi Arabia Education Mission
...,...,...,...,...,...,...,...,...,...,...
28216,28217,4279700,Albert Einstein College of Medicine,Bronx,NY,42704,381717,Contract,CHINA,Chia Tai TianQing Pharmaceutical Gr
28217,28218,4279700,Albert Einstein College of Medicine,Bronx,NY,42778,444938,Contract,ISRAEL,BL Oncology Ltd
28218,28219,4279700,Albert Einstein College of Medicine,Bronx,NY,42907,1064580,Contract,ENGLAND,Roche Products Limited
28219,28220,4279700,Albert Einstein College of Medicine,Bronx,NY,42948,737375,Contract,SWITZERLAND,F Hoffman-La Roche Ltd


--- 

### Describe the Differences between different classes of gift types.



---

### 1) Which country gives the most money in total?

Qatar donated 2706240869, the largest sum by a wide margin. This is compared to runner-up England's 1464906771. 


In [6]:
#sort by country donation, descending

# df.sort_values(by='Foreign Gift Amount', ascending=False)

#group by country and sum the donations
df.groupby('Country of Giftor')['Foreign Gift Amount'].sum().sort_values(ascending=False)


Country of Giftor
QATAR           2706240869
ENGLAND         1464906771
CHINA           1237952112
SAUDI ARABIA    1065205930
BERMUDA          899593972
                   ...    
IRAN                    80
ST. LUCIA               50
CROATIA                 26
SLOVAKIA                20
ERITREA                  3
Name: Foreign Gift Amount, Length: 155, dtype: int64

--- 

### 2) Which country initiates the most gifts by count?

ENGLAND at 3655 total donations. 

In [24]:
df['Country of Giftor'].value_counts().sort_values(ascending=False)

Country of Giftor
ENGLAND        3655
CHINA          2461
CANADA         2344
JAPAN          1896
SWITZERLAND    1676
               ... 
BARBADOS          1
BELARUS           1
ERITREA           1
PALAU             1
ANTIGUA           1
Name: count, Length: 155, dtype: int64

In [8]:
# find rows with country of giftor = ROMANIA (Audit counts - Both check out)
# df[df['Country of Giftor'] == 'ROMANIA']
# df[df['Country of Giftor'] == 'PALAU']

In [9]:
# df

---

### 3) Which country gives the largest gifts on average?

Bermuda gives the largest gifts, with an average donation of just over $7.6M per dontation. 

In [10]:
# Take the average for each country's donation amount, sort by average donation amount, descending
pd.options.display.float_format = '{:,.0f}'.format
df.groupby('Country of Giftor')['Foreign Gift Amount'].mean().sort_values(ascending=False) #.astype(int)


Country of Giftor
BERMUDA     7,688,837
ECUADOR     3,985,140
QATAR       3,905,109
ANTIGUA     3,435,545
JERSEY      3,064,589
               ...   
ST. LUCIA          50
IRAN               40
CROATIA            26
SLOVAKIA           20
ERITREA             3
Name: Foreign Gift Amount, Length: 155, dtype: float64

---

### 4) Which institution receive the most money in total?

Carnegie Mellon University -- Total Donations: $1,477,922,504


In [11]:
df.groupby('Institution Name')['Foreign Gift Amount'].sum().sort_values(ascending=False)

# dfInstitution = df.groupby('Institution Name')
# dfInstitution['Foreign Gift Amount'].sum().sort_values(ascending=False)

Institution Name
Carnegie Mellon University               1477922504
Cornell University                       1289937761
Harvard University                        954803610
Massachusetts Institute of Technology     859071692
Yale University                           613441311
                                            ...    
Bates College                                250000
Metropolitan State University                235439
University of California, Merced             182514
University of Central Oklahoma               160288
University of Jamestown                         500
Name: Foreign Gift Amount, Length: 318, dtype: int64

---

###  5) Which institution receive the greatest number of gifts by count?

University of California, Los Angeles; Total Gifts Received: 3916


In [25]:
df['Institution Name'].value_counts().sort_values(ascending=False)

# dfInstitution['Institution Name'].value_counts().sort_values(ascending=False)

Institution Name
University of California, Los Angeles          3916
California Institute of Technology             2502
Columbia University in the City of New York    1127
Ohio State University (The)                    1014
Johns Hopkins University                        899
                                               ... 
Midway University                                 1
Murray State University                           1
Bates College                                     1
Washington Adventist University                   1
Presidio Graduate School                          1
Name: count, Length: 318, dtype: int64

Build a plot of the top 20 institutions by the amount or number of gifts they have received using either seaborn or plotly. 

In [27]:
# giftPlot = px.bar(df.groupby('Institution Name')['Institution Name'].sort_values(ascending=True).head(20))
giftPlot = px.bar(df['Institution Name'].value_counts().sort_values(ascending=False).head(20))
giftPlot.show()

--- 

### 6) What is the average amount each university receives? Median amount? If they are different why?

By Institution 1) *see plot*
- Median and mean are different because the data is assymetrical. In this case, 'mean of all donations' > 'median of all donations', which indicates that there are large donations contributing to a right-skewed distribution. 



Aggregate 2) 
- Mean accross all Institutions = $588,233
- Median accross all Institutions = $94,615
    - This comparison also illustrates the right skew of the distribution. 



Mean & Median across each *INDIVIDUAL* Institution's donations

In [29]:
# group by insitution, take the average of all their donations, sort by avg amount, descending
df.groupby('Institution Name')['Foreign Gift Amount'].mean().sort_values(ascending=False)

Institution Name
Brigham Young University - Idaho      22,705,516
Western International University      12,942,161
Whittier College                      12,079,000
University of Phoenix                 11,094,346
Curtis Institute of Music             10,000,000
                                         ...    
University of Alabama at Birmingham       29,838
University of California, Merced          15,210
Marquette University                      14,050
Westminster Theological Seminary           7,696
University of Jamestown                      500
Name: Foreign Gift Amount, Length: 318, dtype: float64

In [15]:
# group by insitution, take the median of all their donations, sort by avg amount, descending

df.groupby('Institution Name')['Foreign Gift Amount'].median().sort_values(ascending=False)

Institution Name
Brigham Young University - Idaho        22,794,050
Curtis Institute of Music               10,000,000
Juilliard School (The)                   5,570,000
Indiana State University                 4,291,189
Midwestern State University              4,200,000
                                           ...    
University of Jamestown                        500
University of California, Merced               398
University of California, Los Angeles          250
Claremont McKenna College                      150
Westminster Theological Seminary               100
Name: Foreign Gift Amount, Length: 318, dtype: float64

In [16]:
# donationPlt = px.bar(df.groupby('Institution Name')['Foreign Gift Amount'].mean().sort_values(ascending=False).head(20))



# # add lines for median and mean of all donations, regardless of country or institution
# donationPlt.add_hline(y=df['Foreign Gift Amount'].mean(), line_color="black", annotation_text="Mean of all donations", annotation_position="bottom right")
# donationPlt.add_hline(y=df['Foreign Gift Amount'].median(), line_color="red", annotation_text="Median of all donations", annotation_position="top right")

In [40]:
grouped = (
    df.groupby('Institution Name')['Foreign Gift Amount']
    .mean()
    .sort_values(ascending=False)
    .head(20)
)

donationPlt = px.bar(
    x=grouped.values,
    y=grouped.index,
    orientation='h',
    labels={'x': 'Average Foreign Gift Amount', 'y': 'Institution Name'},
)
fGA = "${:,.0f}".format(df['Foreign Gift Amount'].mean())
mGA = "${:,.0f}".format(df['Foreign Gift Amount'].median())

donationPlt.update_yaxes(autorange="reversed")
donationPlt.add_vline(x=df['Foreign Gift Amount'].mean(), line = dict( color ="black", width = 3), annotation_text= f"Mean of all donations {fGA}", annotation_position="bottom right", annotation_yshift=-30,annotation_xshift = -10)
donationPlt.add_vline(x=df['Foreign Gift Amount'].median(), line = dict(color="red", width = 3) , annotation_text= f"Median of all donations {mGA}", annotation_position="top right", annotation_yshift=20, annotation_xshift = -10, annotation_font_color="red")
donationPlt.show()

--- 

### 7) What is the largest flow of a single country to a single institution?

(Check against alluvial diagram)

Qatar's dontations to Cornell is the largest flow of a single country to a single insitution, totalling $1.02 Billion

In [18]:
#view select rows
pd.set_option('display.max_rows', 15)

# use `pd.crosstab` to look at the relationship between outcoming gifts from countries and incoming gifts to institutions.
# pd.crosstab(df['Country of Giftor'], df['Institution Name'], values=df['Foreign Gift Amount'], aggfunc='sum', margins=True, margins_name="Total Gifts").sort_values(by="Total Gifts", ascending=False).T
#Secondary sorting to better identify biggest gifts, descending
pd.crosstab(df['Country of Giftor'], df['Institution Name'], values=df['Foreign Gift Amount'], aggfunc='sum', margins=True, margins_name="Total Gifts").sort_values(by="Total Gifts", ascending=False).sort_values(by="Total Gifts", axis=1, ascending=False).T



Country of Giftor,Total Gifts,QATAR,ENGLAND,CHINA,SAUDI ARABIA,BERMUDA,CANADA,HONG KONG,JAPAN,SWITZERLAND,...,TOKELAU,ARMENIA,SLOVENIA,BARBADOS,BELARUS,IRAN,ST. LUCIA,CROATIA,SLOVAKIA,ERITREA
Institution Name,,,,,,,,,,,,,,,,,,,,,
Total Gifts,"16,600,515,514","2,706,240,869","1,464,906,771","1,237,952,112","1,065,205,930","899,593,972","898,160,656","887,402,529","655,954,776","619,899,445",...,"6,000",300,115,100,100,80,50,26,20,3
Carnegie Mellon University,"1,477,922,504","425,410,702","4,640,165","20,892,044","1,050,000","750,000,000","13,605,711","13,750,000","30,913,374","8,903,800",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Cornell University,"1,289,937,761","1,018,473,315","12,492,789","27,864,330","2,579,426","980,000","21,730,083","33,440,603","10,192,678","15,000,663",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Harvard University,"954,803,610","5,077,754","188,328,288","87,165,906","28,655,079","4,127,000","31,147,133","140,275,479","17,359,425","59,089,854",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Massachusetts Institute of Technology,"859,071,692",NaN,"70,307,241","47,906,641","89,435,405","5,500,000","5,275,433","65,661,712","46,586,288","7,786,070",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Bates College,"250,000",NaN,"250,000",NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Metropolitan State University,"235,439",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
"University of California, Merced","182,514",NaN,91,NaN,NaN,NaN,602,NaN,"11,500",NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


---

### 8) Which are the top giftors to US academic institutions?

- The top-5 giftors are: 
    - Qatar: 2,706,240,869
    - England: 1,464,906,771
    - China: 1,237,952,112
    - Saudi Arabia: 1,065,205,930
    - Bermuda: 899,593,972

In [19]:
# df.head()

#giftor = 'Giftor Name'
giftor = 'Country of Giftor'
recipi = 'Institution Name'
flow = 'Foreign Gift Amount'
N = 25

flows = (
    df.groupby([giftor, 
                recipi])
      [flow]
      .sum()
      .nlargest(N)
      .reset_index()
)

labels = (
    flows[giftor].tolist()
    + flows[recipi].tolist()
)

labels = list(dict.fromkeys(labels))

fig = go.Figure(
    go.Sankey(
        node=dict(label=labels),
        link=dict(
            source=flows[giftor]
                        .map(labels.index),
            target=flows[recipi]
                        .map(labels.index),
            value=flows[flow]
        )
    )
)

fig.show()